processor/src/host/advice/injectors/adv_stack_injectors.rs

```diff
@@ -296,6 +296,107 @@ pub(crate) fn push_signature<S: ProcessState, A: AdviceProvider>(
     Ok(HostResponse::None)
 }
 
+/// Pushes the number of the leading zeros of the top stack element onto the advice stack.
+///
+/// Inputs:
+///   Operand stack: [n, ...]
+///   Advice stack: [...]
+///
+/// Outputs:
+///   Operand stack: [n, ...]
+///   Advice stack: [leading_zeros, ...]
+pub(crate) fn push_leading_zeros<S: ProcessState, A: AdviceProvider>(
+    advice_provider: &mut A,
+    process: &S,
+) -> Result<HostResponse, ExecutionError> {
+    let n = process.get_stack_item(0).as_int() as u32;
+    let leading_zeros = Felt::new(n.leading_zeros().into());
+    advice_provider.push_stack(AdviceSource::Value(leading_zeros))?;
+    Ok(HostResponse::None)
+}
+
+/// Pushes the number of the trailing zeros of the top stack element onto the advice stack.
+///
+/// Inputs:
+///   Operand stack: [n, ...]
+///   Advice stack: [...]
+///
+/// Outputs:
+///   Operand stack: [n, ...]
+///   Advice stack: [trailing_zeros, ...]
+pub(crate) fn push_trailing_zeros<S: ProcessState, A: AdviceProvider>(
+    advice_provider: &mut A,
+    process: &S,
+) -> Result<HostResponse, ExecutionError> {
+    let n = process.get_stack_item(0).as_int() as u32;
+    let trailing_zeros = Felt::new(n.trailing_zeros().into());
+    advice_provider.push_stack(AdviceSource::Value(trailing_zeros))?;
+    Ok(HostResponse::None)
```

Comment:
This logic is copy/pasted across the other functions. There are a few ways to extract that logic to make it more generally useful, but one option would be something like

```rs
pub(crate) fn push_trailing_zeros<S: ProcessState, A: AdviceProvider>(
    advice_provider: &mut A,
    process: &S,
) -> Result<HostResponse, ExecutionError> {
    push_transformed_stack_top(|stack_top| {
        let stack_top = stack_top.as_int() as u32;
        Felt::new(stack_top.trailing_zeros().into())
    })
}

fn push_transformed_stack_top(f: impl FnOnce(Felt) -> Felt) -> Result<HostResponse, ExecutionError> {
    let stack_top = process.get_stack_item(0);
    let transformed_stack_top = f(stack_top);
    advice_provider.push_stack(AdviceSource::Value(transformed_stack_top))?;
    Ok(HostResponse::None)
}
```

If we also apply the previous comment about returning an error instead of using `as u32`, we make the closure return a `Result<Felt, ExecutionError>` instead of `Felt`